### PR TITLE
With TASK_DIAGNOSTICS enabled, tasks will be automatically profiled

### PR DESCRIPTION
**Pull request by dralley**
[noissue]

### CODE REVIEW

File: pulpcore/tasking/worker.py [hunk 1]
```diff
@@ -471,4 +477,19 @@ def _perform_task(task_pk, task_working_dir_rel_path):
...
-    execute_task(task)
+    if settings.TASK_DIAGNOSTICS:
+        from pyinstrument import Profiler
+        with Profiler() as profiler:
+            execute_task(task)
+
+        with open(diagnostics_dir / "pyinstrument.html", "w+") as f:
+            f.write(profiler.output_html())
+
+        # import cProfile
```
Comment (dralley) [inline pulpcore/tasking/worker.py:488]: pyinstrument pros:

* nicer default output than cprofile, which needs external tools to wrangle
* async-aware
* according to the comparison done by Emery Berger of Scalene fame, more accurate
* less overhead

cProfile pros:

* in the standard library

Comment (dralley) [reply pulpcore/tasking/worker.py:491]: Here's what the Pyinstrument  HTML report looks like when opened in a browser.

Screenshot from 2023-09-15 14-13-03

**Comment by bmbouter** [comment]
:100: A thousand times yes to this idea!

It needs docs too though. I so much enjoy seeing this! Great job.

**Comment by bmbouter** [comment]
Regarding cprofile vs pyinstrument I'm +1 to pyinstrument because we need it to be async aware, and it's just easier to consume the output. We need to make this really easy. I think picking up a dependency (and its dependencies) is worth it.

That's my take, but I'm interested in what others think about that also.

File: pulpcore/tasking/worker.py [hunk 2]
```diff
@@ -471,4 +480,18 @@ def _perform_task(task_pk, task_working_dir_rel_path):
...
+    if settings.TASK_DIAGNOSTICS and has_pyinstrument:
+        from pyinstrument import Profiler
+
+        with Profiler() as profiler:
+            execute_task(task)
+
+        profile_file = diagnostics_dir / "pyinstrument.html"
+        _logger.info("Writing task profile data to {}".format(profile_file))
+        with open(profile_file, "w+") as f:
+            f.write(profiler.output_html())
```
Comment (lubosmj) [inline pulpcore/tasking/worker.py:495]: Does stacking multiple profiles into the same output file really work here?

File: pulpcore/tasking/worker.py [hunk 3]
```diff
@@ -458,9 +463,13 @@ def _perform_task(task_pk, task_working_dir_rel_path):
...
     if settings.TASK_DIAGNOSTICS:
+        diagnostics_dir = task_diagnostics_dir(task_pk)
+        mem_diagnostics_path = diagnostics_dir / "memory.datum"
         # It would be better to have this recording happen in the parent process instead of here
         # https://github.com/pulp/pulpcore/issues/2337
-        normal_thread = threading.Thread(target=write_memory_usage, args=(task_pk,), daemon=True)
+        normal_thread = threading.Thread(
+            target=write_memory_usage, args=(mem_diagnostics_path,), daemon=True
+        )
         normal_thread.start()
```
Comment (mdellweg) [inline pulpcore/tasking/worker.py:473]: What is "normal" about this thread? It is exceptional to run it, and it's marked as daemon. Can we give this a better name?

**Review by lubosmj — APPROVED** [review]

File: pulpcore/tasking/worker.py [hunk 4]
```diff
@@ -471,4 +480,18 @@ def _perform_task(task_pk, task_working_dir_rel_path):
     set_current_user(user)
     set_domain(task.pulp_domain)
     os.chdir(task_working_dir_rel_path)
-    execute_task(task)
+
+    # set up profiling
+    has_pyinstrument = importlib.util.find_spec("pyinstrument") is not None
```
Comment (mdellweg) [inline pulpcore/tasking/worker.py:485]: Hah, I only knew `try except ImportError`.

File: pulpcore/tasking/worker.py (hunk 2 above)
Comment (dralley) [reply pulpcore/tasking/worker.py:495]: What do you mean by multiple profiles?

Comment (lubosmj) [reply pulpcore/tasking/worker.py:495]: The results of each task run are written to the same file. And those results are the "profiles".

Comment (dralley) [reply pulpcore/tasking/worker.py:495]: Should only be one profile per task.  `diagnostics_dir` is `/var/tmp/pulp/<task_id>/`

Comment (ipanova) [reply pulpcore/tasking/worker.py:495]: each task run or other words each task, will have its own pk which is the disagnostics_dir

File: pulpcore/tasking/worker.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulpcore/tasking/worker.py:485]: Do you mean by lazy evaluation? like:
```python
if settings.TASK_DIAGNOSTICS and importlib.util.find_spec("pyinstrument") is not None:
    ....
```
Given we execute these lines with every task, i think we can actually spare us some disc-access here.

File: pulpcore/tasking/worker.py (hunk 2 above)
Comment (dralley) [reply pulpcore/tasking/worker.py:495]: yes

**Review by mdellweg — APPROVED** [review]

**Comment by ipanova** [comment]
@dralley could you ad a line or 2 to the docs https://docs.pulpproject.org/pulpcore/plugin_dev/plugin-writer/concepts/tasks/diagnostics.html#diagnostics

**Pull request merged**